kernel/common/numeric.rb

```diff
@@ -31,23 +31,27 @@ def <=>(other)
     return nil
   end
 
-  def step(limit, step=1)
+  def step(limit=nil, step=nil, to: nil, by: nil)
     unless block_given?
-      return to_enum(:step, limit, step) do
-        Rubinius::Mirror::Numeric.reflect(self).step_size(limit, step)
+      return to_enum(:step, limit, step, to: to, by: by) do
+        Rubinius::Mirror::Numeric.reflect(self).step_size(limit, step, to, by)
       end
     end
 
-    raise ArgumentError, "step cannot be 0" if step == 0
-
     m = Rubinius::Mirror::Numeric.reflect(self)
-    values = m.step_fetch_args(limit, step)
+    values = m.step_fetch_args(limit, step, to, by)
     value = values[0]
     limit = values[1]
     step = values[2]
     asc = values[3]
     is_float = values[4]
 
+    if(step == 0)
+      while true
+        yield self
```

Comment:
At far as I could see, it's not supposed too. It's how MRI behaves... If step, given by the by: keyword argument, is 0, you get an infinite loop.
From the [Numeric#step documentation](http://ruby-doc.org/core-2.1.5/Numeric.html#method-i-step):
> The loop finishes when the value to be passed to the block is greater than limit (if step is positive) or less than limit (if step is negative), where limit is defaulted to infinity.

You can still do stuff like
```
10.step(to: 15, by: 0).take(5)
=> [10, 10, 10, 10, 10]
```
... although I don't immediately see it's usefulness.